### ISSUE TITLE
Unexpected auth state invalidation on multiple tabs/windows with the same client and existing auth state

### ISSUE THREAD

**Issue by bjornharrtell** [adfinis/ember-simple-auth-oidc#760]
After opening a second window of the same client it fails with:

![image](https://user-images.githubusercontent.com/141030/232424896-a1d891dc-0561-4586-bf5f-5324d006f605.png)

**Comment by derrabauke** [comment]
Hi Björn, 
is this happening after an upgrade to the latest version or a brand new setup? If upgraded, try to delete the local storage for your site.

Do you have a `routeAfterAuthentication` defined in your [config]?

**Comment by bjornharrtell** [comment]
@derrabauke thanks for the pointers. Deleting local storage has no effect and I don't `routeAfterAuthentication` defined. It's an upgraded older application so looks like I will have to try and reproduce with a more bare bones new client to rule out any code smell.

**Comment by bjornharrtell** [comment]
Oh I now understand you as I should have `routeAfterAuthentication` defined! However, I find it strange that it would not also trigger the problem on a single instance of the client.

**Comment by bjornharrtell** [comment]
Setting `routeAfterAuthentication` makes it go past the routing error but it doesn't resolve the problem - it instead exposes it as it looks like it passes "already authenticated" but then proceeds to get unauthorized from the backend. So looks like the process invalidates the backend auth session somehow... and `authenticated` property in local storage does end up as an empty object. It seems likely there is something not quite right either in my backend or frontend code... Hopefully I'll be able to figure it out and report findings back here.

**Comment by bjornharrtell** [comment]
It's taking some time to setup a bare bones project.. but plot thickens in that I find that the second client window gives the following debug output:

DEBUG: The authenticator "authenticator:oidc" rejected to restore the session - invalidating…
DEBUG: Error: Refresh token is missing

**Comment by bjornharrtell** [comment]
I've now managed to create a bare bones from scratch ember client using ember-cli 4.12.1 and ember-simple-auth-oidc 5.0.3 and get the same problem so I'm starting to think this is due to some detail in the backend auth/idp implementation vs. ember-simple-auth-oidc.

**Comment by bjornharrtell** [comment]
@derrabauke with help from an associate we've come to the conclusion that the behaviour is to be expected with pkce and auth info saved in local storage (and hence shared between tabs for the same domain).

**Comment by bjornharrtell** [comment]
I jumped the gun there.. it's expected when the authentication flow is run again, it should be possible to reuse existing session from local storage. As I can reproduce in minimal example it might be a problem in ember-simple-auth-oidc in that it initate auth flow in this case?

**Comment by derrabauke** [comment]
> I've now managed to create a bare bones from scratch ember client using ember-cli 4.12.1 and ember-simple-auth-oidc 5.0.3 and get the same problem so I'm starting to think this is due to some detail in the backend auth/idp implementation vs. ember-simple-auth-oidc.

Re-running the auth flow on the second tab should not happen if it has sufficient session data in the local storage. Can you see, if there are requests happening which ends in a `401` response? (Maybe you need to tick *Preserve logs* in the network tab.) 

I will jump onto it and check what's happening. On our apps opening a second window/tab seems to work as intended. :thinking: I assume, you've tested it in different browsers, already!?

**Comment by bjornharrtell** [comment]
I had up to this point just tested with Chrome. No 401 responses are seen.. and it does seem like I get valid token etc from the oauth backend but authentication in local storage is still ending up cleared after the second tab has gone through redirect back to the client.

But to add to the confusion... I just tried the same in Firefox and I cannot reproduce the issue (!). It does look like it goes throught he auth flow and redirect back from the auth backend every time though.. but end result is working regardless.

**Comment by derrabauke** [comment]
> I had up to this point just tested with Chrome. No 401 responses are seen.. and it does seem like I get valid token etc from the oauth backend but authentication in local storage is still ending up cleared after the second tab has gone through redirect back to the client.
> 
> But to add to the confusion... I just tried the same in Firefox and I cannot reproduce the issue (!). It does look like it goes through the auth flow and redirect back from the auth backend every time though.. but end result is working regardless.

So your problem is resolved? :sparkles:

**Comment by bjornharrtell** [comment]
@derrabauke afraid not.. must have it work in Chrome (if it wasn't clear above - it was just observations on behaviour the issue is the same). And I want to figure out why it does full auth flow, that doesn't seem right even if it does not fail in Firefox.

**Comment by bjornharrtell** [comment]
I've now set my bare bones example up against auth0 (previous attempts use another product) but issue is the same, so I'm inclined to rule out the auth backend as the cause.

**Comment by bjornharrtell** [comment]
@derrabauke looks like we could use some serious help here before it becomes a critical issue, do you have time and interest if we invest into this in some way?

**Comment by derrabauke** [comment]
> @derrabauke looks like we could use some serious help here before it becomes a critical issue, do you have time and interest if we invest into this in some way?

hey Björn, first it seemed as if this was an issue with your auth-backend, but now with the repro on auth0, the suspicion confirmed that something is wrong within our implementation. Could you share the repro with us, so we can try to debug it?

Cheers

**Comment by bjornharrtell** [comment]
@derrabauke yep pushed it at https://github.com/bjornharrtell/ember-test-oicd, just need to supply an auth0 endpoint and clientId.

When I have it running I can visit https://localhost:4200/test fine in a single tab but if I try to open that same URL in a second tab it ends up at / and empty "authenticated" in local storage.

It also seems to me it does full auth flow also just reloading when opened in a single tab which even though functional it's not desireable since it causes longer startup time and burden on the auth backend.

**Comment by derrabauke** [comment]
Hey @bjornharrtell; I checked out your repo and could reproduce the behavior. Somehow the local storage stays empty. I will further debug it tomorrow.

**Comment by bjornharrtell** [comment]
@derrabauke thanks for the status update and that sounds promising! Let me know if there anything we can do to help, including sponsorship of some sort.

**Comment by derrabauke** [comment]
> @derrabauke thanks for the status update and that sounds promising! Let me know if there anything we can do to help, including sponsorship of some sort.

hi @bjornharrtell - We found a solution! 

**TL;DR**: add `offline_access` to your `scope` config. Like this:
```js
'ember-simple-auth-oidc': {
      host: 'https://XXXX.auth0.com',
      scope: 'openid, offline_access', // <--- HERE
      clientId: 'XXX',
      authEndpoint: '/authorize',
      tokenEndpoint: '/oauth/token',
      endSessionEndpoint: '/endsession',
      userinfoEndpoint: '/userinfo',
      afterLogoutUri: '/',
      enablePkce: true,
    },

```
We will definitely add a hint in the *README* and eventually add the `offline_access` scope by default to the `scope` configuration key. But for the second part we first have to check for further implications.

---

**Digging deep** into the OIDC specifications, we can find a [specifiaction for *Offline Access*](https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess) right before the *Refresh Token* Chapters, which states: 
> The use of Refresh Tokens is not exclusive to the offline_access use case. The Authorization Server MAY grant Refresh Tokens in other contexts that are beyond the scope of this specification.

This sounds really vague, and gives a lot of room of interpretation for implementations of the standard.

Further the [OAuth 2.0 Authorizatio Framework](https://www.rfc-editor.org/rfc/rfc6749#section-1.5) defines *refresh tokens* as optional:
> Issuing a refresh token is optional at the discretion of the
   authorization server.  If the authorization server issues a refresh
   token, it is included when issuing an access token.

That's why it was not an issue in our environments yet, since we mostly use Keycloak and this provides us a *refresh token* by default. But *auth0* on the other hand, does not provide a *refresh token* by default. There you would have to add the `offline_access` scope as [described here](https://auth0.com/docs/secure/tokens/refresh-tokens).

Other `oidc` plugins have had similar discussions, as you can see in the example of [this angular oidc](https://github.com/manfredsteyer/angular-oauth2-oidc/issues/1241) package.

When enabling the *refresh token* you should always consider a sensible setting for *refresh token rotation*. Don't let it life for to long. You can find a good read on that [in the oauth.com blog](https://www.oauth.com/oauth2-servers/making-authenticated-requests/refreshing-an-access-token/).

---
 
 **Regarding your offer of a sponsorship:** There is a initiative running, which is quite important for the ember eco-system. [It's organized by a company named Mainmatter, which is gathering sponsorships for financing a dev team to get the ember build system to the next level](https://twitter.com/marcoow/status/1651119303775625218). Maybe you could think about a sponsorship this instead. :) If you'd like more info on that I can send you a brochure. Just reach out to me at falk.neumann@adfinis.com .

**Comment by bjornharrtell** [comment]
@derrabauke thanks! I'm yet to understand the finer details of offline_access and the relation to this problem but it does seem to do the trick which I guess is the important bit at this point. :)
